### PR TITLE
Adds analytics for Login/Scope

### PR DESCRIPTION
**Pull request by discorick**
This should help us understand better if there are any issues/objections in our login funnel when users attempt to authorize via GitHub

### CODE REVIEW

File: vendor/engines/saas/lib/saas/engine.rb [hunk 1]
```diff
@@ -17,6 +17,19 @@ class Engine < ::Rails::Engine
         })
       end
 
+      LoginController.class_eval do
+        before_action :page_job, only: [:public, :private]
+
+        :private
+
+        def page_job
+          if request.referer !~ /github\.com/
```
Comment (dahlbyk) [inline vendor/engines/saas/lib/saas/engine.rb]: This seems kind of brittle - what if we checked for one of the OAuth parameters?

**Comment by dahlbyk** [comment]
This looks pretty good, will kick the tires on the review app after lunch and hopefully :ship: :it:.

File: vendor/engines/saas/lib/saas/engine.rb (hunk 1 above)
Comment (discorick) [inline vendor/engines/saas/lib/saas/engine.rb]: I attempted to filter out by OAuth params `code` & `state`, but theres actually 3 calls on `/login/scope` during the process of authentication - 2 of which have no OAuth params which creates duplicate pages.

**Comment by dahlbyk** [comment]
I'm seeing a triple-tap on `/login/{scope}` in Segment, all from Ruby.
image

**Comment by dahlbyk** [comment]
> I'm seeing a triple-tap on `/login/{scope}` in Segment, all from Ruby.

To be more specific, it only seems to triple-tap if I've already authorized GitHub. From a clean/revoked state the event count is correct.

**Comment by discorick** [comment]
Ah, ill address that good catch

**Comment by discorick** [comment]
@dahlbyk , I can deal with the triple tap easily enough - having a double is a lot more difficult - I can't find any distinguishing request data between the initial login and the initial redirect. Any thoughts ?

**Comment by dahlbyk** [comment]
> I can't find any distinguishing request data between the initial login and the initial redirect check.

Didn't quite parse that. Can you be more specific?

**Comment by discorick** [comment]
In the situation where you are already authorized, on `/login/private` the page is called once, then the controller action executes up to `github_authenticate! :private` . Once warden is satisfied the token is good, the controller then redirects to `login/private` again. I am just looking for a differentiator for this second request to filter the page out

```ruby
 def private
    request.env['warden'].logout if github_authenticated? :default
    github_authenticate! :private # <<<< Right Here
    @user = gh.user
    @emails = @user.emails.all
    redirect_to params[:redirect_to] || "/"
  end
```

File: vendor/engines/saas/lib/saas/engine.rb [hunk 2]
```diff
@@ -17,6 +17,20 @@ class Engine < ::Rails::Engine
...
       end
 
+      LoginController.class_eval do
+        before_action :page_job, only: [:public, :private]
+
+        :private
+
+        def page_job
+          auth_request = request.params['code'] && request.params['state']
+          if request.referer !~ /github\.com/ && !logged_in? && !auth_request
```
Comment (dahlbyk) [inline vendor/engines/saas/lib/saas/engine.rb]: Am I right in thinking the `request.referer` check is redundant given the other two?

1. `/login`
2. `/login/{scope}` (not `logged_in?` and `!auth_request`, so page is logged)
3. https://github.com/login/oauth/authorize
4. `/login/{scope}?code=...&state=...` (`auth_request` is true)
5. `/login/{scope}` (`logged_in?` should now be true)

Comment (discorick) [inline vendor/engines/saas/lib/saas/engine.rb]: True of theres already a token, referrer is needed for a login that is not yet authorized with GitHub, as the code and state are not yet available in request #2.

Comment (dahlbyk) [inline vendor/engines/saas/lib/saas/engine.rb]: Right, but this particularly code is mean to log `/login/{scope}` only at step 2. Step 5 handles the post-auth tracking of `/login/{scope}/authorized`, unless I'm missing something?

Comment (discorick) [inline vendor/engines/saas/lib/saas/engine.rb]: Yes, bouncing the referrer on an unauthenticated login prevents a step #4 duplicate Page (`state` and `code`) aren't part of the params just yet in this scenario.

Comment (dahlbyk) [inline vendor/engines/saas/lib/saas/engine.rb]: :OK: let's leave it. This is in the SaaS engine so Enterprise should not be affected, correct?

**Comment by dahlbyk** [comment]
Am I missing something or did the latest here not get deployed?

File: vendor/engines/saas/lib/saas/engine.rb (hunk 2 above)
Comment (discorick) [inline vendor/engines/saas/lib/saas/engine.rb]: Correct

**Comment by discorick** [comment]
@dahlbyk Not sure what you mean ?

**Comment by dahlbyk** [comment]
I don't see a deploy notification after the commit you pushed most recently.

**Comment by dahlbyk** [comment]
> I don't see a deploy notification after the commit you pushed most recently.

Confirmed. 08e9657 was not deployed for some reason.
firefox_2016-03-12_11-27-08

Recreating Review App.

**Comment by dahlbyk** [comment]
This is looking good.

One edge case we're not handling:

1. Revoke app
2. Login with public access
3. Hit `/repositories/private/{login}`
4. Approve extended permissions (N.B. pleased to see the `APP_NAME` redirect works in this scenario)
5. Auth succeeds without being tracked

This is probably somewhat rare, but if we wanted to test https://github.com/huboard/huboard-web/pull/181 we'd want to be able to tell when private auth succeeds. Any ideas?

**Comment by discorick** [comment]
Another good catch @dahlbyk , looks like logins via `/repositories/(scope)/login` are handled via the `DashboardsController`

**Comment by discorick** [comment]
@dahlbyk, just looked into this in more detail. 

I think for the time being it might be worth leaving this pathway un-tracked as `/login/private/authorized` - whats actually happening is a scope change, and complimentary to #181 I think we should release analytics surrounding the upgrading/downgrading of scopes in the improved login flow.

**Comment by dahlbyk** [comment]
> I think we should release analytics surrounding the upgrading/downgrading of scopes in the improved login flow.

:+1: Merged as-is.

Also related to https://github.com/huboard/huboard-web/issues/172#issuecomment-196328505

**Comment by rauhryan** [comment]
nice 👍

**Pull request merged**